app/services/authentication/authenticator.rb

```diff
@@ -0,0 +1,127 @@
+module Authentication
+  # Authenticator will perform one of these tree operations:
+  # 1. create a new user and match it to its authentication identity
+  # 1. update an existing user and align it to its authentication identity
+  # 1. return the current user if a user is given (already logged in scenario)
+  class Authenticator
+    # auth_payload is the payload schema, see https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema
+    def initialize(auth_payload, current_user: nil, cta_variant: nil)
+      @provider = load_authentication_provider(auth_payload)
+
+      @current_user = current_user
+      @cta_variant = cta_variant
+    end
+
+    def self.call(*args)
+      new(*args).call
+    end
+
+    def call
```

Comment:
the main logic here is unchanged. this works both with existing users and new users in a bunch of scenarios, all tested in the corresponding spec